proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,392 @@
+   when we select the heaviest fork in the next step, we should see all the
+   potential candidates for optimistically confirmed slots, there we can count
+   the votes and remove some false positive cases.
+
+   However, it's also overkill to repair every block presented by others. When
+   `RestartLastVotedForkSlots` messages are being received and aggregated, a
+   validator can categorize blocks missing locally into 2 categories: must-have
+   and ignored.
+
+   We set the line at 42% when 80% join the restart, it's possible that
+   different validators see different 80%, so their must-have blocks might
+   be different, but in reality this case should be rare. Whenever some block
+   gets to 42%, repair could be started, because when more validators join the
+   restart, this number will only go up but will never go down.
+
+   Once the validator gets `RestartLastVotedForkSlots`, it can calculate which
+   blocks must be repaired. When all those "must-have" blocks are repaired and
+   replayed, it can proceed to step 3.
+
+3. **gossip current heaviest fork**
+
+   The main goal of this step is to "vote" the heaviest fork to restart from.
+
+   We use a new gossip message `RestartHeaviestFork`, its fields are:
+
+   * `slot`: `u64` slot of the picked block.
+   * `hash`: `Hash` bank hash of the picked block.
+   * `stake_committed_percent`: `u16` total percentage of stakes of the
+   validators it received `RestartHeaviestFork` messages from.
+
+   After receiving `RestartLastVotedForkSlots` from the validators holding
+   stake more than `RESTART_STAKE_THRESHOLD` and repairing slots in "must-have"
+   category, replay all blocks and pick the heaviest fork like this:
+
+   1. Calculate the threshold for a block to be on the heaviest fork, the
+   heaviest fork should have all blocks with possibility to be optimistically
+   confirmed. The number is `67% - 5% - stake_on_validators_not_in_restart`.
+
+   For example, if 80% validators are in restart, the number would be
+   `67% - 5% - (100-80)% = 42%`. If 90% validators are in restart, the number
+   would be `67% - 5% - (100-90)% = 52%`.
+
+   2. Sort all blocks passing the calculated threshold, and verify that they
+   form a single chain. If it's the first block in the list, its parent block
+   should be the local root. Otherwise its parent block should be the one
+   immediately ahead of it in the list.
+
+   If any block does not satisfy above constraint, print the first offending
+   block and exit.
+
+   3. If the list is empty, then output local root as the HeaviestFork.
+   Otherwise output the last block in the list as the HeavistFork.
+
+   To see why the above algorithm is safe, we will prove that:
+
+   1. Any block optimistically confirmed before the restart will always be
+   on the list:
+
+   Assume block A is one such block, it would have `67%` stake, discounting
+   `5%` non-conforming and people not participating in wen_restart, it should
```

Comment:
Because validators can misbehave either due to being malicious or due to software bugs, we want to use non-conforming which doesn't automatically imply people are behaving poorly.